rst_files/wald_friedman.rst

```diff
@@ -401,467 +345,201 @@ that appear on the right side of the Bellman equation as separate functions
-        ub = pgrid[searchsortedlast(J - p_c_0, -1e-10)]
-
-        return lb, ub
-    end
-
-    function solve_model!(wf; tol = 1e-7)
-        bell_op(x) = bellman_operator(wf, x)
-        J =  compute_fixed_point(bell_op, zeros(wf.m), err_tol=tol, print_skip=5)
-
-        wf.sol.J = J
-        wf.sol.lb, wf.sol.ub = find_cutoff_rule(wf, J)
-        return J
-    end
-
-    function simulate(wf, f; p0 = 0.5)
-        # Check whether vf is computed
-        if sum(abs, wf.sol.J) < 1e-8
-            solve_model!(wf)
+        roots = find_zeros(p -> cost(p, d0, d1, L0, L1, c), 0, 1 - eps())
+        left = first.(choice.(roots .- eps(), d0, d1, L0, L1, c))
+        right = first.(choice.(roots .+ eps(), d0, d1, L0, L1, c))
+        β, α = (0.0, 1.0)
+        for (left, right, root) in zip(left, right, roots)
+            if left == 2 && right ≠ 2
+                β = root
+            end
         end
-
-        # Unpack useful info
-        lb, ub = wf.sol.lb, wf.sol.ub
-        drv = DiscreteRV(f)
-
-        # Initialize a couple useful variables
-        decision = 0
-        p = p0
-        t = 0
-
-        while true
-            # Maybe should specify which distribution is correct one so that
-            # the draws come from the "right" distribution
-            k = rand(drv)
-            t = t + 1
-            p = bayes_update_k(wf, p, k)
-            if p < lb
-                decision = 1
-                break
-            elseif p > ub
-                decision = 0
-                break
+        for (left, right, root) in zip(reverse(left), reverse(right), reverse(roots))
+            if left ≠ 1 && right == 1
+                α = root
             end
         end
-
-        return decision, p, t
-    end
-
-    abstract type HiddenDistribution end
-    struct F0 <: HiddenDistribution end
-    struct F1 <: HiddenDistribution end
-
-    function simulate_tdgp(wf, f; p0 = 0.5)
-        decision, p, t = simulate(wf, wf.f0; p0=p0)
-
-        correct = (decision == 0)
-
-        return correct, p, t
+        if β < α
+            @printf("Accept x1 if p ≤ %.2f\nContinue to draw if %.2f ≤ p ≤ %.2f\nAccept x0 if p ≥ %.2f", β, β, α, α)
+        else
+            x0 = accept_x0(β, L0)
+            x1 = accept_x1(β, L1)
+            draw = draw_again(β, d0, d1, L0, L1, c, min(x0, x1))
+            if draw == min(x0, x1, draw)
+                @printf("Accept x1 if p ≤ %.2f\nContinue to draw if %.2f ≤ p ≤ %.2f\nAccept x0 if p ≥ %.2f", β, β, α, α)
+            else
+                @printf("Accept x1 if p ≤ %.2f\nAccept x0 if p ≥ %.2f", β, α)
+            end
+        end
+        return (α, β)
     end
 
-    function simulate_tdgp(wf, f; p0 = 0.5)
-        decision, p, t = simulate(wf, wf.f1; p0=p0)
+We can simulate an agent facing a problem and the outcome with the following function
 
-        correct = (decision == 1)
+.. code-block:: julia
 
-        return correct, p, t
+    @with_kw struct Problem
```

Comment:
Is there a reason we're not using a `NamedTuple`? Like: 

```
Problem = @with_kw (d0 = Beta(1,1), d1 = Beta(9,9), L0 = 2, L1 = 2, c = 0.2, p = 0.5, n = 100, return_output = false) # Constructor for problem objects. 
DefaultProblem = Problem()
ModifiedProblem = Problem(d0 = Beta(1, 3))
```

Reply:
Yes: https://github.com/QuantEcon/lecture-source-jl/blob/master/style.md#type-annotations-parameters-and-generic-programming

We want to avoid these kinds of structs at all costs.  Not to mention that this struct is not templatized for the distributions, so it would be holding abstract types.

Luckily, named tuples do it all.

Reply:
Good catch. I have made the changes to use `NamedTuple`.